lib/image-editor/class-image-editor.php

```diff
@@ -0,0 +1,260 @@
+		$target_file = $this->get_filename( $info['meta'] );
+
+		// Does the image already exist?
+		$existing_image = $this->get_existing_image( $info, $target_file );
+		if ( $existing_image ) {
+			// Return the existing image.
+			return $existing_image;
+		}
+
+		// Try and load the image itself.
+		$image = $this->load_image( $media_id, $info );
+		if ( is_wp_error( $image ) ) {
+			return $image;
+		}
+
+		foreach ( $modifiers as $modifier ) {
+			// Finally apply the modification.
+			$modified = $modifier->apply_to_image( $image['editor'] );
+			if ( is_wp_error( $modified ) ) {
+				return $modified;
+			}
+		}
+
+		// And save.
+		return $this->save_image( $image, $target_file, $info );
+	}
+
+	/**
+	 * Loads an image for editing.
+	 *
+	 * @param integer $media_id Image ID.
+	 * @return array|WP_Error The WP_Image_Editor and image path if successful, WP_Error otherwise.
+	 */
+	private function load_image( $media_id ) {
+		require_once ABSPATH . 'wp-admin/includes/image.php';
+
+		$image_path = get_attached_file( $media_id );
+
+		if ( empty( $image_path ) ) {
+			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		$image_editor = wp_get_image_editor( $image_path );
+		if ( ! $image_editor->load() ) {
+			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		return array(
+			'editor' => $image_editor,
+			'path'   => $image_path,
+		);
+	}
+
+	/**
+	 * Gets the JSON response object for an image.
+	 *
+	 * @param integer $id Image ID.
+	 * @return array Image JSON.
+	 */
+	private function get_image_as_json( $id ) {
+		$path     = '/wp/v2/media/' . $id;
+		$response = rest_do_request( $path );
+
+		if ( ! $response->is_error() ) {
+			$response->set_status( 201 );
+			$response->header( 'Location', rest_url( $path ) );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Checks for the existence of an image and if it exists, return the image.
+	 *
+	 * @param array  $attachment Attachment with url to look up.
+	 * @param string $target_file Target file name to look up.
+	 * @return array|false Image JSON if exists, otherwise false.
+	 */
+	private function get_existing_image( $attachment, $target_file ) {
+		$url = str_replace( basename( $attachment['url'] ), $target_file, $attachment['url'] );
+
+		$new_id = attachment_url_to_postid( $url );
+		if ( $new_id > 0 ) {
+			return $this->get_image_as_json( $new_id );
+		}
+
+		return false;
+	}
+
+	/**
+	 * Saves an edited image.
+	 *
+	 * @param array  $image_edit Image path and editor to save.
+	 * @param string $target_name Target file name to save as.
+	 * @param array  $attachment Attachment with metadata to apply.
+	 * @return array|WP_Error Image JSON if successful, WP_Error otherwise
+	 */
+	private function save_image( $image_edit, $target_name, $attachment ) {
```

Comment:
To confirm, this creates a new attachment? Are existing alt, description and caption fields copied from the original image?